onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
     return true;
   }
 
+  bool are_input_same = inputs_.size() == other.inputs_.size();
+  
+  if(are_input_same) {
+    for(unsigned long i=0; i<inputs_.size(); i++) {
+      std::vector<const EquivalenceClass*> iclasses_ = inputs_[i];
+      std::vector<const EquivalenceClass*> oclasses_ = other.inputs_[i];
+      
+      are_input_same = are_input_same && (iclasses_.size() == oclasses_.size());
+      
+      if(!are_input_same)
+        break;
+      
+      for(unsigned long j=0; j<iclasses_.size(); j++) {
+        const EquivalenceClass* clsi = iclasses_[j];
+        const EquivalenceClass* clso = oclasses_[j];
+        
+        if((clsi->initializer_ != nullptr) && (clso->initializer_ != nullptr)) {
```

Comment:
If we compare initializers in a pairwise manner we can end up omitting nodes that _can potentially be part of the same equivalence class_, say for example a elementwise Mul operation `Mul(x, 2) == Mul(2, x)` where x is some tensor. 

We should have some mechanism to handle nodes that follow commutative property - will check in a test case for the same. What do you think @yuslepukhin ?

Reply:
Test case:
```python3
import torch
import torch.nn.functional as F
import numpy.random as npr

class TestCSE(torch.nn.Module):
  def __init__(self):
    super(TestCSE, self).__init__()
    self.outputs = []

  def forward(self, x):
    mult_const = torch.from_numpy(npr.rand(1,3,5,5))
    for i in range(0, 5):
      weights = torch.from_numpy(npr.rand(1,3,3,3))

      if i%2 ==0:
        y = mult_const * x
      else:
        y = x * mult_const
      # even and odd convolutions will have same pad
      # so all even and odd convolution;s parent pad
      # must belong to the same EquivalenceClass
      padded_x  = F.pad(y, (i%2 + 1, i%2 + 1, i%2 + 1, i%2 + 1))
      out = F.conv2d(padded_x, weights)
      self.outputs.append(out)

    return self.outputs

torch.onnx.export(TestCSE(), torch.from_numpy(npr.rand(1, 3, 5, 5)), "ex3.onnx", verbose=True)
```
Input net - 

<img width="710" alt="Screenshot 2020-10-28 at 12 53 10 PM" src="https://user-images.githubusercontent.com/4116675/97404726-8b31af80-191c-11eb-9807-9e172fdd9be6.png">

Output net - 

<img width="731" alt="Screenshot 2020-10-28 at 12 53 33 PM" src="https://user-images.githubusercontent.com/4116675/97404761-97b60800-191c-11eb-982b-100e837a8db3.png">